python/dgl/graphbolt/item_sampler.py

```diff
@@ -510,68 +313,94 @@ def __init__(
+                offsets.append(offsets[-1] + len(value[0]))
+            else:
+                raise TypeError(
+                    f"Unsupported buffer type {type(value).__name__}."
                 )
-                for key in keys
-            }
-        return default_collate(batch)
+        return torch.tensor(offsets)
 
     def __iter__(self) -> Iterator:
-        if self._use_indexing:
-            seed = self._rng.integers(0, np.iinfo(np.int32).max)
-            data_pipe = IterableWrapper(
-                ItemShufflerAndBatcher(
-                    self._item_set,
-                    self._shuffle,
-                    self._batch_size,
-                    self._drop_last,
-                    self._buffer_size,
-                    distributed=self._distributed,
-                    drop_uneven_inputs=self._drop_uneven_inputs,
-                    world_size=self._world_size,
-                    rank=self._rank,
-                    rng=np.random.default_rng(seed),
-                )
-            )
+        worker_info = torch.utils.data.get_worker_info()
+        if worker_info is not None:
+            num_workers = worker_info.num_workers
+            worker_id = worker_info.id
         else:
-            # Organize items.
-            data_pipe = self._organize_items(self._item_set)
-
-            # Collate.
-            data_pipe = data_pipe.collate(collate_fn=self._collate)
-
-        # Map to minibatch.
-        data_pipe = data_pipe.map(partial(self._minibatcher, names=self._names))
+            num_workers = 1
+            worker_id = 0
+        total = len(self._item_set)
+        start_offset, assigned_count, output_count = calculate_range(
+            self._distributed,
+            total,
+            self._world_size,
+            self._rank,
+            num_workers,
+            worker_id,
+            self._batch_size,
+            self._drop_last,
+            self._drop_uneven_inputs,
+        )
+        if self._shuffle:
+            g = torch.Generator()
+            g.manual_seed(self._seed + self._epoch)
+            _permutation = torch.randperm(total, generator=g)
+            buffer = self._item_set[
+                _permutation[start_offset : start_offset + assigned_count]
+            ]
+        else:
+            buffer = self._item_set[
+                start_offset : start_offset + assigned_count
+            ]
+        offsets = self._calculate_offsets(buffer)
+        for i in range(0, assigned_count, self._batch_size):
+            if output_count <= 0:
+                break
+            indices = torch.arange(i, i + min(self._batch_size, output_count))
+            output_count -= self._batch_size
+            yield self._minibatcher(
+                self._collate_batch(buffer, indices, offsets),
+                self._names,
+            )
 
-        return iter(data_pipe)
+        self._epoch += 1
 
 
 class DistributedItemSampler(ItemSampler):
```

Comment:
does it work well for `num_workers>0`?

Reply:
No, the issue #7381 still exist. I'll continue to investigate.